generic/tests/ioquit.py

```diff
@@ -42,16 +43,20 @@ def run(test, params, env):
     session2.cmd(check_cmd, timeout=360)
 
     error_context.context("Kill the VM", logging.info)
-    vm.process.close()
-
+    try:
+        process.kill_process_tree(vm.process.get_pid(), wait=30)
```

Comment:
Tested with \"process.kill_process_tree\" 50 times, the results are all failed.  Qemu is alive. 
And with vm.process.close(), the results are all passed.

some logs:
00:34:41 INFO | Context: Kill the VM
00:34:41 WARNI| Older Avocado without support to wait for killed pids, waiting for 10s but it still might be alive
00:34:51 INFO | Context: Check img after kill VM
00:34:51 DEBUG| Checking image file /home/kvm_autotest_root/images/rhel76-64-virtio-scsi.qcow2
00:34:51 INFO | Running \'/bin/qemu-img info /home/kvm_autotest_root/images/rhel76-64-virtio-scsi.qcow2 --output=human\'
00:34:51 DEBUG| [stderr] qemu-img: Could not open \'/home/kvm_autotest_root/images/rhel76-64-virtio-scsi.qcow2\': Failed to get shared \"write\" lock
00:34:51 DEBUG| [stderr] Is another process using the image?

Reply:
This is because of 2 issues. First you haven't applied the https://github.com/avocado-framework/avocado/pull/2924 which is not yet part of Avocado and is the proper fix (together with this patch) and second because I forgot to kill the machine in case that feature is missing (backward compatibility with older avocado).

Anyway updated the code so even the fallback path should work well.

Btw this is a real issue, on ppc64le with xfs in ~50% of cases we're hitting this issue and I verified it's because the shell spawned by aexpect is dead, but the qemu process is still alive.

Reply:
Applied avocado-framework/avocado#2924, and retested on rhel7.6. Test passed but there were warnings during execution.
logs:
......
05:46:42 WARNI| Older Avocado without support to wait for killed pids, waiting for 10s but it still might be alive
05:46:42 INFO | Running \'kill -19 15134\'
......
......

Reply:
PS: I'm sorry for not testing the fallback branch properly... I only run 13 jobs with the avocado fix:

```
avocado run ioquit --vt-guest-os RHEL.7.devel -m /tmp/a.yaml                                                               
JOB ID     : 3f19b07620cc92d10b344d90c7ade5c41de5db6a
JOB LOG    : /root/avocado/job-results/job-2018-11-28T15.13-3f19b07/job.log                                                                         
 (01/13) io-github-autotest-qemu.ioquit;foo-9860: PASS (141.52 s)
 (02/13) io-github-autotest-qemu.ioquit;bar-b171: PASS (129.56 s)
 (03/13) io-github-autotest-qemu.ioquit;baz-278c: PASS (156.73 s)
 (04/13) io-github-autotest-qemu.ioquit;1-8e3f: PASS (150.44 s)
 (05/13) io-github-autotest-qemu.ioquit;2-a70f: PASS (144.14 s)
 (06/13) io-github-autotest-qemu.ioquit;3-5109: PASS (166.77 s)
 (07/13) io-github-autotest-qemu.ioquit;4-de69: PASS (143.75 s)
 (08/13) io-github-autotest-qemu.ioquit;5-dbfe: PASS (161.31 s)
 (09/13) io-github-autotest-qemu.ioquit;6-71e6: PASS (126.37 s)
 (10/13) io-github-autotest-qemu.ioquit;7-c658: PASS (160.43 s)
 (11/13) io-github-autotest-qemu.ioquit;8-90e6: PASS (134.06 s)
 (12/13) io-github-autotest-qemu.ioquit;9-b351: PASS (126.64 s)
 (13/13) io-github-autotest-qemu.ioquit;0-c47d: PASS (151.22 s)
RESULTS    : PASS 13 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0                                                                  
JOB TIME   : 1897.81 s
JOB HTML   : /root/avocado/job-results/job-2018-11-28T15.13-3f19b07/results.html                                                                    
```

Those are results with the fix but without avocado part:

```
avocado run ioquit --vt-guest-os RHEL.7.devel -m /tmp/a.yaml 
JOB ID     : 3921242aafaae3d9b0b0e01fcc947d8cfe647005
JOB LOG    : /root/avocado/job-results/job-2018-11-29T04.42-3921242/job.log
 (01/13) io-github-autotest-qemu.ioquit;foo-9860: WARN: Test passed but there were warnings during execution. Check the log for details. (137.01 s)
 (02/13) io-github-autotest-qemu.ioquit;bar-b171: WARN: Test passed but there were warnings during execution. Check the log for details. (133.49 s)
 (03/13) io-github-autotest-qemu.ioquit;baz-278c: WARN: Test passed but there were warnings during execution. Check the log for details. (130.12 s)
 (04/13) io-github-autotest-qemu.ioquit;1-8e3f: WARN: Test passed but there were warnings during execution. Check the log for details. (135.19 s)
 (05/13) io-github-autotest-qemu.ioquit;2-a70f: WARN: Test passed but there were warnings during execution. Check the log for details. (118.55 s)
 (06/13) io-github-autotest-qemu.ioquit;3-5109: /^C
Interrupt requested. Waiting 2 seconds for test to finish (ignoring new Ctrl+C until then)
...
```

where the warning informs about older avocado and potential issues, but none of the 5 tests failed.

The original results (without any fixes):

```
avocado run ioquit --vt-guest-os RHEL.7.devel -m /tmp/a.yaml                                                               
JOB ID     : 481c2226cb93580327f4eb73015383e8174f89e9
JOB LOG    : /root/avocado/job-results/job-2018-11-29T05.14-481c222/job.log                                                                         
 (01/13) io-github-autotest-qemu.ioquit;foo-9860: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (142.82 s)
 (02/13) io-github-autotest-qemu.ioquit;bar-b171: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (124.42 s)
 (03/13) io-github-autotest-qemu.ioquit;baz-278c: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (146.68 s)
 (04/13) io-github-autotest-qemu.ioquit;1-8e3f: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (150.82 s)
 (05/13) io-github-autotest-qemu.ioquit;2-a70f: PASS (128.33 s)
 (06/13) io-github-autotest-qemu.ioquit;3-5109: PASS (102.00 s)
 (07/13) io-github-autotest-qemu.ioquit;4-de69: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (99.50 s)
 (08/13) io-github-autotest-qemu.ioquit;5-dbfe: PASS (109.18 s)
 (09/13) io-github-autotest-qemu.ioquit;6-71e6: PASS (115.48 s)
 (10/13) io-github-autotest-qemu.ioquit;7-c658: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (112.06 s)
 (11/13) io-github-autotest-qemu.ioquit;8-90e6: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (120.51 s)
 (12/13) io-github-autotest-qemu.ioquit;9-b351: PASS (99.08 s)
 (13/13) io-github-autotest-qemu.ioquit;0-c47d: WARN: qemu-img check exceptions. Some bad data in the image may have gone unnoticed (/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2) (102.11 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 8 | INTERRUPT 0 | CANCEL 0                                                                   
JOB TIME   : 1558.04 s
JOB HTML   : /root/avocado/job-results/job-2018-11-29T05.14-481c222/results.html                                                                    
```

Where the qemu-img exception reports:

```
2018-11-29 05:40:01,235 process          L0448 DEBUG| [stderr] qemu-img: Could not open '/var/lib/libvirt/images/avocado/avocado-vt/images/rhel7devel-ppc64.qcow2': Failed to get shared "write" lock
```

Reply:
@XueqiangWei the warning is there only when your Avocado does not support the `wait` argument, therefor without the applied patch. Have you `install/deploy` the Avocado after applying the patch?